aten/src/ATen/core/Generator.h

```diff
@@ -1,24 +1,62 @@
 #pragma once
 
-#include <ATen/core/ATenGeneral.h>
 #include <stdint.h>
+#include <memory>
+#include <mutex>
+#include <deque>
+#include <atomic>
+#include <typeinfo>
+#include <utility>
+
+#include <c10/util/Exception.h>
+#include <c10/util/C++17.h>
+#include <ATen/Device.h>
+
+/*
+* Generator note.
+* A Pseudo Random Number Generator (PRNG) is an engine that uses an algorithm to 
+* generate a seemingly random sequence of numbers, that may be later be used in creating 
+* a random distribution. Such an engine almost always maintains a state and requires a
+* seed to start off the creation of random numbers. Often times, users have
+* encountered that it could be beneficial to be able to create, retain, and destroy 
+* PRNG states and also be able to have control over the seed value.
+*
+* A Generator in ATen gives users the ability to read, write and modify a PRNG engine.
+* For instance, it does so by letting users seed a PRNG engine, fork the state of the
+* engine, etc.
+*
+* By default, there is one generator per device, and a device's generator is 
+* lazily created. A user can use the torch.Generator() api to create their own generator.
+* Currently torch.Generator() can only create a CPUGenerator.
+*/
 
 namespace at {
 
+constexpr uint64_t default_rng_seed_val = 67280421310721;
+
 struct CAFFE2_API Generator {
-  Generator() {};
-  Generator(const Generator& other) = delete;
-  Generator(Generator&& other) = delete;
-  virtual ~Generator() {};
-
-  virtual Generator& copy(const Generator& other) = 0;
-  virtual Generator& free() = 0;
-
-  virtual uint64_t seed() = 0;
-  virtual uint64_t initialSeed() = 0;
-  virtual Generator& manualSeed(uint64_t seed) = 0;
-  virtual Generator& manualSeedAll(uint64_t seed) = 0;
-  virtual void * unsafeGetTH() = 0;
+  // Constructors
+  Generator(Device device_in, uint64_t seed_in);
+  Generator(const Generator& other);
+  Generator(Generator&& other);
+  virtual ~Generator() = default;
+
+  // Common methods for all generators
+  virtual void setCurrentSeed(uint64_t seed);
+  uint64_t getCurrentSeed();
+  Device getDevice();
+
+  // stubbed. will be removed
+  virtual Generator& manualSeedAll(uint64_t seed);
+
+  protected:
+    mutable std::mutex mutex;
+    Device device_;
+    uint64_t current_seed_;
```

Comment:
I think we need to think more clearly about what thread-safety guarantees we want this class to give. This is important because you've made `device_` and `current_seed_` protected variables, which means that people who write subclasses can access these directly (without going through the locked getter); if you don't give clear guidance about thread safety expectations, they will probably get it wrong.

My understanding (which could be wrong) is that the most important thing to make thread safe is the random number generator state itself; changes to device or current seed don't actually happen concurrently from different threads.

Reply:
That is an important point about protected variables that I completely overlooked! You are right that the most important thing to lock is the rng. I see this pattern that people like to call `torch.manual_seed()` before they use their rand functions. So I thought if that would raise thread safety issue if I don't look the getters and setters as well.

Reply:
I feel like usually manual seed is called at the beginning of the program, or from a single threaded context. So that's why I hypothesize it's OK not to lock. Although it's going to setup the RNG state, which we should lock just for good reason, so maybe not.

TBH, the fact that we are sharing RNG state across threads is kind of crazy, it really should be a thread local concept. But I agree this is consistent with what we did previously, so...